### ISSUE TITLE
PANIC at dmu.c:1016:dmu_write()

### ISSUE THREAD

**Issue by stefarossi** [zfsonlinux/zfs#6622]
I have a problem with a pool, consisting of a single 8TB drive.
I think the problem originated from a software issue rather than a hardware one, as I can't find anything weird in the SMART log.
The drive was attached to a Mac running [O3X](https://openzfsonosx.org). One morning when I woke up I found that the Mac had restarted due to a kernel panic. My pool had been exported, and my other pool (two 3TB drives in mirroring mode) was in a degraded state with one of the mirrors marked as faulty. I was able to import the 8TB pool again with `zpool import -fFX`, but then the system started panicking. I was only able to solve this by detaching the drive.

I now moved all my drives to a Debian box. I fixed the 3TB pool by re-adding the mirror, but I can't fix the 8TB one. I can only import it with `-o readonly=on`, but one of the three dataset remains inaccessible:
```
unable to fetch ZFS version for filesystem 'Pool3/DatasetName'
cannot mount 'Pool3/DatasetName': Resource temporarily unavailable
```
`zpool status -v` shows this:
 ```
  pool: Pool3
 state: ONLINE
status: One or more devices has experienced an error resulting in data
	corruption.  Applications may be affected.
action: Restore the file in question if possible.  Otherwise restore the
	entire pool from backup.
   see: http://zfsonlinux.org/msg/ZFS-8000-8A
  scan: none requested
config:

	NAME        STATE     READ WRITE CKSUM
	Pool3       ONLINE       0     0     1
	  sdb       ONLINE       0     0     4

errors: Permanent errors have been detected in the following files:

        <metadata>:<0x2c>
        <metadata>:<0x62>
        Pool3/DatasetName:<0x0>
```

If I try importing the pool in RW mode, I get this:
```
Message from syslogd@ hostname at Sep  9 20:53:23 ...

 kernel:[75891.300835] VERIFY3(0 == dmu_buf_hold_array(os, object, offset, size, 0, ((char *)__func__), &numbufs, &dbp)) failed (0 == 5)

Message from syslogd@hostname at Sep  9 20:53:23 ...

 kernel:[75891.300926] PANIC at dmu.c:1016:dmu_write()
```

and, in /var/log/messages:
```
Sep  9 20:53:23 puck kernel: [75891.300964] CPU: 1 PID: 26095 Comm: txg_sync Tainted: P           O    4.9.0-3-amd64 #1 Debian 4.9.30-2+deb9u3
Sep  9 20:53:23 puck kernel: [75891.300967] Hardware name: HP ProLiant MicroServer, BIOS O41     10/01/2013
Sep  9 20:53:23 puck kernel: [75891.300971]  0000000000000000 ffffffffaef28574 ffffffffc11eecc4 ffffa633c162fad0
Sep  9 20:53:23 puck kernel: [75891.300979]  ffffffffc06e23a5 0000000000000286 0000000000000030 ffffa633c162fae0
Sep  9 20:53:23 puck kernel: [75891.300986]  ffffa633c162fa80 2833594649524556 756d64203d3d2030 6c6f685f6675625f
Sep  9 20:53:23 puck kernel: [75891.300992] Call Trace:
Sep  9 20:53:23 puck kernel: [75891.301006]  [<ffffffffaef28574>] ? dump_stack+0x5c/0x78
Sep  9 20:53:23 puck kernel: [75891.301032]  [<ffffffffc06e23a5>] ? spl_panic+0xc5/0x100 [spl]
Sep  9 20:53:23 puck kernel: [75891.301174]  [<ffffffffc10b99e0>] ? dmu_buf_hold_array_by_dnode+0x3b0/0x470 [zfs]
Sep  9 20:53:23 puck kernel: [75891.301291]  [<ffffffffc10cd81e>] ? dnode_rele_and_unlock+0x4e/0x80 [zfs]
Sep  9 20:53:23 puck kernel: [75891.301402]  [<ffffffffc10b9f90>] ? dmu_buf_hold_array.constprop.12+0x80/0xb0 [zfs]
Sep  9 20:53:23 puck kernel: [75891.301513]  [<ffffffffc10ba17f>] ? dmu_write.part.7+0xef/0x100 [zfs]
Sep  9 20:53:23 puck kernel: [75891.301644]  [<ffffffffc1110bd4>] ? space_map_write+0x314/0x560 [zfs]
Sep  9 20:53:23 puck kernel: [75891.301660]  [<ffffffffc06dd08a>] ? spl_kmem_zalloc+0x8a/0x160 [spl]
Sep  9 20:53:23 puck kernel: [75891.301666]  [<ffffffffaecec192>] ? getrawmonotonic64+0x32/0xc0
Sep  9 20:53:23 puck kernel: [75891.301791]  [<ffffffffc10f0ecd>] ? metaslab_sync+0x14d/0x7a0 [zfs]
Sep  9 20:53:23 puck kernel: [75891.301901]  [<ffffffffc10b7347>] ? ddt_sync+0x337/0x620 [zfs]
Sep  9 20:53:23 puck kernel: [75891.302023]  [<ffffffffc10ea29a>] ? dsl_scan_sync+0x4a/0xc40 [zfs]
Sep  9 20:53:23 puck kernel: [75891.302129]  [<ffffffffc10ac0ff>] ? bplist_iterate+0xaf/0x100 [zfs]
Sep  9 20:53:23 puck kernel: [75891.302260]  [<ffffffffc11176bd>] ? vdev_sync+0x5d/0x120 [zfs]
Sep  9 20:53:23 puck kernel: [75891.302389]  [<ffffffffc1100738>] ? spa_sync+0x468/0xd80 [zfs]
Sep  9 20:53:23 puck kernel: [75891.302520]  [<ffffffffc1112ad1>] ? txg_sync_thread+0x2e1/0x4b0 [zfs]
Sep  9 20:53:23 puck kernel: [75891.302652]  [<ffffffffc11127f0>] ? txg_delay+0x1a0/0x1a0 [zfs]
Sep  9 20:53:23 puck kernel: [75891.302667]  [<ffffffffc06df230>] ? __thread_exit+0x20/0x20 [spl]
Sep  9 20:53:23 puck kernel: [75891.302682]  [<ffffffffc06df29d>] ? thread_generic_wrapper+0x6d/0x80 [spl]
Sep  9 20:53:23 puck kernel: [75891.302689]  [<ffffffffaec965d7>] ? kthread+0xd7/0xf0
Sep  9 20:53:23 puck kernel: [75891.302694]  [<ffffffffaec96500>] ? kthread_park+0x60/0x60
Sep  9 20:53:23 puck kernel: [75891.302699]  [<ffffffffaec96500>] ? kthread_park+0x60/0x60
Sep  9 20:53:23 puck kernel: [75891.302705]  [<ffffffffaf2064f5>] ? ret_from_fork+0x25/0x30
```

I tried the FreeBSD live cd, but it too crashes and reboots.
Is there any way to recover that dataset?

System information:
```
Type                         | Version/Name
---                          | --- 
Distribution Name            | Debian
Distribution Version         | 9
Linux Kernel                 | 4.9.0-3
Architecture                 | amd64
ZFS Version                  | 0.7.1-1
SPL Version                  | 0.7.1-1
```

Thanks a lot.

**Comment by bunder2015** [comment]
Duplicate of #4830

**Comment by stefarossi** [comment]
You're right, that's the same issue. Sorry for the duplicate.
So, is my only option to clone my data to a new drive and to try importing it?

**Comment by gmelikov** [comment]
Closed as duplicate of #4830.
